File: src/main/java/org/mitallast/queue/raft/LogIndexMap.java
package org.mitallast.queue.raft;

import gnu.trove.list.array.TLongArrayList;
import gnu.trove.map.TObjectLongMap;
import gnu.trove.map.hash.TObjectLongHashMap;
import javaslang.collection.Set;
import org.mitallast.queue.raft.cluster.ClusterConfiguration;
import org.mitallast.queue.raft.cluster.JointConsensusClusterConfiguration;
import org.mitallast.queue.transport.DiscoveryNode;

public class LogIndexMap {
    private final TObjectLongMap<DiscoveryNode> backing;
    private final TLongArrayList indexes = new TLongArrayList(64);

    public LogIndexMap(long defaultIndex) {
        this.backing = new TObjectLongHashMap<>(64, 0.5f, defaultIndex);
    }

    public long decrementFor(DiscoveryNode member) {
        long value = indexFor(member) - 1;
        backing.put(member, value);
        return value;
    }

    public void put(DiscoveryNode member, long value) {
        backing.put(member, value);
    }

    public long putIfGreater(DiscoveryNode member, long value) {
        if (backing.containsKey(member)) {
            long prev = backing.get(member);
            if (prev < value) {
                backing.put(member, value);
                return value;
            } else {
                return prev;
            }
        } else {
            backing.put(member, value);
            return value;
        }
    }

    public long consensusForIndex(ClusterConfiguration config) {
        if (config.isTransitioning()) { // joint
            long oldQuorum = indexOnMajority(((JointConsensusClusterConfiguration) config).getOldMembers());
            long newQuorum = indexOnMajority(((JointConsensusClusterConfiguration) config).getNewMembers());
            return Math.min(oldQuorum, newQuorum);
        } else { // stable
            return indexOnMajority(config.members());
        }
    }

    private long indexOnMajority(Set<DiscoveryNode> include) {
        if (include.isEmpty()) {
            return 0;
        }
        indexes.resetQuick();
        include.forEach(node -> indexes.add(indexFor(node)));
        indexes.sort();
        int index = ceiling(include.size(), 2) - 1;
        return indexes.get(index);
    }

    private int ceiling(int numerator, int divisor) {
        if (numerator % divisor == 0) {
            return numerator / divisor;
        } else {
            return (numerator / divisor) + 1;
        }
    }

    public long indexFor(DiscoveryNode member) {
        return backing.get(member);
    }
}

File: src/main/java/org/mitallast/queue/raft/LogIndexMap.kt
package org.mitallast.queue.raft

import gnu.trove.list.array.TLongArrayList
import gnu.trove.map.TObjectLongMap
import gnu.trove.map.hash.TObjectLongHashMap
import io.vavr.collection.Set
import org.mitallast.queue.raft.cluster.ClusterConfiguration
import org.mitallast.queue.raft.cluster.JointConsensusClusterConfiguration
import org.mitallast.queue.transport.DiscoveryNode

class LogIndexMap(defaultIndex: Long) {
    private val backing: TObjectLongMap<DiscoveryNode>
    private val indexes = TLongArrayList(64)

    init {
        this.backing = TObjectLongHashMap(64, 0.5f, defaultIndex)
    }

    fun decrementFor(member: DiscoveryNode): Long {
        val value = indexFor(member) - 1
        backing.put(member, value)
        return value
    }

    fun put(member: DiscoveryNode, value: Long) {
        backing.put(member, value)
    }

    fun putIfGreater(member: DiscoveryNode, value: Long): Long {
        if (backing.containsKey(member)) {
            val prev = backing.get(member)
            if (prev < value) {
                backing.put(member, value)
                return value
            } else {
                return prev
            }
        } else {
            backing.put(member, value)
            return value
        }
    }

    fun consensusForIndex(config: ClusterConfiguration): Long {
        return when(config) {
            is JointConsensusClusterConfiguration -> {
                val oldQuorum = indexOnMajority(config.oldMembers)
                val newQuorum = indexOnMajority(config.newMembers)
                Math.min(oldQuorum, newQuorum)
            }
            else -> {
                indexOnMajority(config.members)
            }
        }
    }

    private fun indexOnMajority(include: Set<DiscoveryNode>): Long {
        if (include.isEmpty) {
            return 0
        }
        indexes.resetQuick()
        include.forEach { node -> indexes.add(indexFor(node)) }
        indexes.sort()
        val index = ceiling(include.size(), 2) - 1
        return indexes.get(index)
    }

    private fun ceiling(numerator: Int, divisor: Int): Int {
        return if (numerator % divisor == 0) {
            numerator / divisor
        } else {
            numerator / divisor + 1
        }
    }

    fun indexFor(member: DiscoveryNode): Long {
        return backing.get(member)
    }
}